node/src/components/network.rs

```diff
@@ -0,0 +1,610 @@
+        if env::var(ENABLE_LIBP2P_ENV_VAR).is_err() {
+            let network = Network {
+                our_id,
+                peers: HashSet::new(),
+                listening_addresses: vec![],
+                known_addresses,
+                shutdown_sender: Some(server_shutdown_sender),
+                server_join_handle: None,
+                _phantom_data: PhantomData,
+            };
+            return Ok((network, Effects::new()));
+        }
+
+        if notify {
+            debug!("our node id: {}", our_id);
+        }
+
+        // Create a keypair for authenticated encryption of the transport.
+        let noise_keys = noise::Keypair::<X25519Spec>::new()
+            .into_authentic(&id_keys)
+            .map_err(Error::StaticKeypairSigning)?;
+
+        // Create a tokio-based TCP transport.  Use `noise` for authenticated encryption and `yamux`
+        // for multiplexing of substreams on a TCP stream.
+        let transport = TokioTcpConfig::new()
+            .nodelay(true)
+            .upgrade(upgrade::Version::V1)
+            .authenticate(NoiseConfig::xx(noise_keys).into_authenticated())
+            .multiplex(YamuxConfig::default())
+            .timeout(CONNECTION_TIMEOUT)
+            .boxed();
+
+        // Create a Swarm to manage peers and events.
+        let mut swarm = {
+            let ping = Ping::new(PingConfig::new().with_keep_alive(true));
+            let behavior = Behavior {
+                our_id: our_id.clone(),
+                ping,
+            };
+
+            SwarmBuilder::new(transport, behavior, our_peer_id)
+                .executor(Box::new(|future| {
+                    tokio::spawn(future);
+                }))
+                .build()
+        };
+
+        // Schedule connection attempts to known peers.
+        for address in known_addresses.keys() {
+            Swarm::dial_addr(&mut swarm, address.clone()).map_err(|error| Error::DialPeer {
+                address: address.clone(),
+                error,
+            })?;
+        }
+
+        // Specify listener.
+        let listening_address = address_str_to_multiaddr(config.bind_address.as_str());
+        Swarm::listen_on(&mut swarm, listening_address.clone()).map_err(|error| Error::Listen {
+            address: listening_address,
+            error,
+        })?;
+
+        // Start the server task.
+        let server_join_handle = Some(tokio::spawn(server_task(
+            event_queue,
+            server_shutdown_receiver,
+            our_id.clone(),
+            swarm,
+        )));
+
+        let network = Network {
+            our_id,
+            peers: HashSet::new(),
+            listening_addresses: vec![],
+            known_addresses,
+            shutdown_sender: Some(server_shutdown_sender),
+            server_join_handle,
+            _phantom_data: PhantomData,
+        };
+        Ok((network, Effects::new()))
+    }
+
+    fn handle_connection_established(
```

Comment:
Should we maybe remove all tracking of peers from this component and delegate that to libp2p?

Reply:
I think we'll need to track peers ourself in order to send targeted messages, but certainly until we replace our gossiper component we will.

Reply:
I _assumed_ that using a multiaddr with a peer ID is a possibility?

Reply:
I think a multiaddr can hold the peer ID as a component of the address.  Not sure I see how that relates to your original question though?

So, if we want to let libp2p handle this and remove the `self.peers` member, I think we'd either need a couple of channels to ask the swarm for the peer list and to respond, or we'd keep a mutexed handle to the swarm, neither of which appeals much to me.  Do you have a better idea?  (I do like the notion of not duplicating the swarm's tracking of peers - I'm just not seeing a good way to do it)

Reply:
We'll experiment and follow up later then, I guess.